### ISSUE TITLE
Memory leaks in Lua - Calling from C

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#1891]
I just came across this blog post:

http://henryla.logdown.com/posts/83526-memory-leaks-in-lua-calling-from-c

Is this something I need to keep in mind with Hammerspoon as well @asmagill & @cmsj ?

**Comment by latenitefilms** [comment]
Related: http://lua-users.org/wiki/WeakTablesTutorial

**Comment by latenitefilms** [comment]
I guess my question is, for code such as this:

```
local function timeOptions()
    local timeOptionsTable = {}
    for i=1, 10 do
        timeOptionsTable[#timeOptionsTable + 1] = {
            label = i18n(string.lower(tools.numberToWord(i))) .. " " .. i18n("secs", {count=i}),
            value = i,
        }
    end
    return timeOptionsTable
end
```

Should I be making `timeOptionsTable` a weak table by adding `setmetatable(timeOptionsTable, { __mode = 'v' })`?

**Comment by asmagill** [comment]
Weakness only applies to objects (tables, userdata, etc.) and not values or strings (technically strings are objects, but for this purpose, they are considered values -- see [2.5.2](http://www.lua.org/manual/5.3/manual.html#2.5.2) of the lua manual, specifically:

>Only objects that have an explicit construction are removed from weak tables. Values, such as numbers and light C functions, are not subject to garbage collection, and therefore are not removed from weak tables (unless their associated values are collected). Although strings are subject to garbage collection, they do not have an explicit construction, and therefore are not removed from weak tables.

If you make `timeOptionsTable` have weak values, you'll have your entries disappear on you when garbage collection occurs, for example:

~~~lua
> a = setmetatable({ { b = 1 }, 2, { 34 } }, { __mode="v" })


> inspect(a)
{ {
    b = 1
  }, 2, { 34 } }

> collectgarbage()
0

> inspect(a)
{
  [2] = 2
}
~~~

Notice that the table values of `a` disappear after garbage collection occurs -- that's because they were tables (objects) that had no other reference than `a` which holds its values weakly.

In the example shown by the person you reference in your first post of this thread, they weren't doing an actual collection cycle (`collectgarbage("count")` doesn't *do* collection, it just returns a measure of memory in use) -- but with the for/do loop, lua had many steps between removing the first object and the last, so internal collection likely had happened at least once, thus the lower count.

Basically, never trust "count" unless you've called `collectgarbage()` at least twice before hand, and even then depending upon how busy your Hammerspoon application is, that will only give you an approximation.  If you know you've got a lot of memory to free up that you'll no longer need, then you can call `collectgarbage()` yourself; otherwise the interpreter will do so eventually.  Do so twice because if the object has a finalizer (a `__gc` function attached to it), this might not get called until the second invocation (thus preventing the object from being truly freed until the second call).

**Comment by latenitefilms** [comment]
Legend, thanks so much @asmagill !! HUGELY appreciated.
